### PR TITLE
Polymorphism Part II

### PR DESCRIPTION
**Pull request by kunjsong01**
This patch solves multi-level polymorphism, enabled 10 more test cases.
Pass rate:  **24/43** TCs passed in `esbmc-cpp/polymorphism_bringup` suite. 

### Dev notes:
- add a loop in `clang_cpp_convert_vft` to create multiple thunk functions to deal with multi-level polymorhism
- fixed conditions in `perform_virtual_dispatch` to make sure the function call really needs dynamic dispatching:
-- excluded ctor
-- excluded dot operator
-- now using `performsVirtualDispatch` 

This patch enabled 10 more test cases (TCs): 
5_MI_Slvl
5_MI_Slvl_fail
9_MI_Mlvl_ndiamond_mcol_2ndcolnval
9_MI_Mlvl_ndiamond_mcol_2ndcolnval_fail
10_MI_Mlvl_ndiamond_mcol_2ndcolval
10_MI_Mlvl_ndiamond_mcol_2ndcolval_fail
llbmc_multiple_inheritance
vmd_defined_inside_class
vmd_defined_inside_class_fail
polymorphism_common_error_5

### Changes to Regression Test Suite:
Added new TCs: 
11_MI_Mlvl_ndiamond_scol
11_MI_Mlvl_ndiamond_scol_fail
vmd_defined_outside_class
vmd_defined_outside_class_fail
vmd_defined_inside_class
vmd_defined_inside_class_fail
polymorphism_common_error_5

### CODE REVIEW

File: src/clang-c-frontend/clang_c_adjust_expr.cpp [hunk 1]
```diff
@@ -522,7 +522,7 @@ void clang_c_adjust::adjust_type(typet &type)
 
     if(s == nullptr)
     {
-      log_error("type symbol `{}' not found", identifier);
+      log_error("type symbol `{}' not found in {}", identifier, __func__);
```
Comment (fbrausse) [inline src/clang-c-frontend/clang_c_adjust_expr.cpp]: Could you put the source location info in front of the actual message in order to clarify that it's not part of that message itself, e.g.
```c++
log_error("{}: type symbol `{}' not found", __func__, identifier);
```

**Comment by fbrausse** [comment]
>     * add a loop in `clang_cpp_convert_vft` to create multiple thunk functions to deal with multi-level polymorhism

Where can I find this loop? :)

File: src/clang-cpp-frontend/clang_cpp_convert_bind.cpp [hunk 2]
```diff
@@ -28,31 +28,47 @@ CC_DIAGNOSTIC_POP()
...
   {
   // TODO: dtor might be virtual too
   //case clang::Decl::CXXDestructor:
   case clang::Decl::CXXMethod:
   {
+    const clang::CXXMethodDecl &cxxmd =
+      static_cast<const clang::CXXMethodDecl &>(decl);
+
+    clang::LangOptions langOpts;
+    langOpts.CPlusPlus = 1;
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert_bind.cpp:50]: Why do you assign `1` here?

File: src/clang-cpp-frontend/clang_cpp_convert_bind.cpp [hunk 3]
```diff
@@ -28,31 +28,47 @@ CC_DIAGNOSTIC_POP()
...
+    langOpts.RTTI = 1;
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert_bind.cpp:51]: Why do you assign `1` here?

**Review by lucasccordeiro — APPROVED** [review]
Minor comments only.

File: src/clang-c-frontend/clang_c_adjust_expr.cpp (hunk 1 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_adjust_expr.cpp]: Done.

File: src/clang-cpp-frontend/clang_cpp_convert_bind.cpp (hunk 2 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert_bind.cpp:50]: To set LangOptions in C++ mode with RTTI enabled.

Added this explanation in the code as a comment.

File: src/clang-cpp-frontend/clang_cpp_convert_bind.cpp (hunk 3 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert_bind.cpp:51]: To set LangOptions in C++ mode with RTTI enabled.

Added this explanation in the code as a comment.

**Comment by lucasccordeiro** [comment]
Many thanks for this PR, @kunjsong01!

**Pull request merged**